src/misc/degree_sequence.cpp

```diff
@@ -68,123 +70,282 @@ template<typename T> inline bool degree_less(const T &a, const T &b) {
+                                            igraph_realize_degseq_t method) {
+    igraph_integer_t n_nodes = igraph_vector_int_size(degseq);
 
-        vd_pair vd = *pt;
-        vertices.erase(pt);
+    // ----- upfront error/graphicality checks ----- //
+    if (n_nodes == 0 || (n_nodes == 1 && VECTOR(*degseq)[0] == 0)) {
+        return IGRAPH_SUCCESS;
+    }
 
-        if (vd.degree == 0) {
-            continue;
+    for (igraph_integer_t i = 0; i < n_nodes; i++) {
+        igraph_integer_t deg = VECTOR(*degseq)[i];
+        if (deg >= n_nodes) {
+            IGRAPH_ERROR("The given degree sequence cannot be realized as a simple graph.", IGRAPH_EINVAL);
         }
+    }
 
-        igraph_integer_t k;
-        vlist::iterator it;
-        for (it = vertices.begin(), k = 0;
-             k != vd.degree && it != vertices.end();
-             ++it, ++k) {
-            if (--(it->degree) < 0) {
-                goto fail;
-            }
-
-            VECTOR(*edges)[2 * (ec + k)] = vd.vertex;
-            VECTOR(*edges)[2 * (ec + k) + 1] = it->vertex;
+    // ----- main Havel-Hakimi loop ----- //
+    // O(V + alpha(V) * E)
+    // O(V + E) for the LARGEST_FIRST method
+    igraph_vector_int_t seq;
+    IGRAPH_CHECK(igraph_vector_int_init_copy(&seq, degseq));
+    IGRAPH_FINALLY(igraph_vector_int_destroy, &seq);
+
+    HavelHakimiList vault(&seq);
+
+    igraph_integer_t n_edges_added = 0;
+    for (igraph_integer_t i = 0; i < n_nodes; i++) {
+        // hub node selection
+        vd_pair hub;
+        if (method == IGRAPH_REALIZE_DEGSEQ_SMALLEST) {
+            if (!vault.get_min_node(/* out param */hub)) break;
+            vault.remove_min_node();
         }
-        if (it == vertices.end() && k < vd.degree) {
-            goto fail;
+        else if (method == IGRAPH_REALIZE_DEGSEQ_LARGEST) {
+            if (!vault.get_max_node(/* out param */hub)) break;
+            vault.remove_max_node();
         }
-
-        ec += vd.degree;
+        else if (method == IGRAPH_REALIZE_DEGSEQ_INDEX) {
+            igraph_integer_t degree = VECTOR(seq)[i];
+            hub = vd_pair{i, degree};
+            vault.buckets[degree].count--;
+        }
+        VECTOR(seq)[hub.vertex] = 0;
+
+        // spoke nodes selection
+        igraph_vector_int_t spokes;
+        IGRAPH_VECTOR_INT_INIT_FINALLY(&spokes, 0);
+        IGRAPH_CHECK(igraph_vector_int_reserve(&spokes, hub.degree));
+        IGRAPH_CHECK(vault.get_spokes(hub.degree, seq, spokes));
+
+        igraph_integer_t n_spokes = igraph_vector_int_size(&spokes);
+        for (igraph_integer_t j = 0; j < n_spokes; j++) {
+            igraph_integer_t spoke_idx = VECTOR(spokes)[j];
+            VECTOR(*edges)[2*n_edges_added] = hub.vertex;
+            VECTOR(*edges)[2*n_edges_added + 1] = spoke_idx;
+            n_edges_added++;
+
+            VECTOR(seq)[spoke_idx]--;
+        }
+        IGRAPH_FINALLY_CLEAN(1);
     }
-
+    IGRAPH_FINALLY_CLEAN(1);
```

Comment:
I see two FINALLY_CLEANs here, both of which seem to be missing a preceding destructor.